<launch>
    <!-- Arguments -->
    <arg name="mode" default="sitl"/>
    <arg name="ns_prefix" default="uav_"/>
    <arg name="robot_model" default="mbzirc"/>
    <arg name="pub_rate" default="50.0"/>
    <arg name="debug" default="false"/>
    <arg name="save_test_data" default="false"/>
    <arg name="save_experiment_data" default="false"/>
    <arg name="trajectory" default="false"/>
    <arg name="reach_tolerance" default="0.1"/>
    <arg name="path" default="basket"/>
    <arg name="generator_mode" default="0"/>
    <arg name="pkg_name" default="upat_follower"/>
    <arg name="use_class" default="true"/>
    <arg name="multi" default="false"/>

    <node pkg="rviz" type="rviz" name="rviz_node" args="-d $(find upat_follower)/config/rviz/mision.rviz" required="true"/>

    <group if="$(eval mode=='sitl')">
        <!-- Launch Gazebo simulation -->
        <rosparam param="/use_sim_time">true</rosparam>
        <node pkg="px4_bringup" type="launch_gzworld.py" name="gazebo_world" output="screen" args="-physics=ode -world=$(find obstacle_avoidance)/config/world/basket_court.world -add_model_path=$(find obstacle_avoidance)/config/models">
            <rosparam param="sim_origin">[37.41036, -6.002353, 6.48]</rosparam>
        </node>
    </group>

    <!-- Launch robot id=1 -->
    <include file="$(find px4_bringup)/launch/spawn_robot.launch" ns="$(arg ns_prefix)1">
        <arg name="robot_model" value="$(arg robot_model)"/>
        <arg name="id" value="1"/>
        <arg name="material" value="Yellow"/>
        <arg name="mode" value="$(arg mode)"/>
        <arg name="x" value="0"/>
        <arg name="y" value="0"/>
        <arg name="z" value="0"/>
        <arg name="yaw" value="0"/>
        <!-- 1.57 -->
    </include>

    <!-- Launch UAL server for drone id=1 -->
    <node pkg="uav_abstraction_layer" type="server" name="ual" output="screen" ns="$(arg ns_prefix)1">
        <param name="uav_id" value="1"/>
        <rosparam param="home_pose">[37.41036, -6.002353, 6.48]</rosparam>        <!-- [x,y,z] -->
    </node>

    <!-- UAV Path Manager nodes -->
    <group ns="upat_follower">
        <group ns="$(arg ns_prefix)1">
            <node pkg="upat_follower" type="ual_communication_node" name="ual_communication" output="screen" required="true">
                <param name="uav_id" value="1"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="save_test_data" value="false"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="path" value="$(arg path)"/>
                <param name="reach_tolerance" value="$(arg reach_tolerance)"/>
                <param name="use_class" value="true"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
                <param name="debug" value="$(arg debug)"/>
            </node>
            <node pkg="upat_follower" type="visualization_node" name="visualization" required="true" output="screen">
                <param name="uav_id" value="1"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="robot_model" value="$(arg robot_model)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="save_experiment_data" value="$(arg save_experiment_data)"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
            </node>
        </group>
    </group>

</launch>
